src/java/org/jivesoftware/util/cache/CacheFactory.java

```diff
@@ -623,38 +611,72 @@ public static synchronized void initialize() throws InitializationException {
         	 throw new InitializationException(e);
         }
     }
-
-    private static ClassLoader getClusteredCacheStrategyClassLoader() {
-        PluginManager pluginManager = XMPPServer.getInstance().getPluginManager();
-        Plugin plugin = pluginManager.getPlugin("hazelcast");
-        if (plugin == null) {
-            plugin = pluginManager.getPlugin("clustering");
-            if (plugin == null) {
-                plugin = pluginManager.getPlugin("enterprise");
-            }
-        }
-        PluginClassLoader pluginLoader = pluginManager.getPluginClassloader(plugin);
-        if (pluginLoader != null) {
-        	if (log.isDebugEnabled()) {
-        		StringBuffer pluginLoaderDetails = new StringBuffer("Clustering plugin class loader: ");
-        		pluginLoaderDetails.append(pluginLoader.getClass().getName());
-        		for (URL url : pluginLoader.getURLs()) {
-        			pluginLoaderDetails.append("\n\t").append(url.toExternalForm());
-        		}
-        		log.debug(pluginLoaderDetails.toString());
-        	}
-            return pluginLoader;
-        }
-        else {
-            log.warn("CacheFactory - Unable to find a Plugin that provides clustering support.");
-            return Thread.currentThread().getContextClassLoader();
-        }
+    
+    /**
+     * Returns the cluster support plugin names
+     * 
+     * @return cluster support plugin names
+     */
+    public static List<String> getClusterSupportPluginNames(){
+    	// list all plugin and check the clusteredCacheFactoryClass implement class
+    	List<String> result = new ArrayList<String>();
+    	PluginManager pluginManager = XMPPServer.getInstance().getPluginManager();
+    	Collection<Plugin> plugins = pluginManager.getPlugins();
+    	for (Plugin plugin : plugins) {
+    		// get the classloader from plugin
+    		PluginClassLoader pluginClassloader = pluginManager.getPluginClassloader(plugin);
+    		try {
+				pluginClassloader.loadClass(clusteredCacheFactoryClass).newInstance();
+				result.add(pluginManager.getName(plugin));
+			} catch (Exception e) {
+				// ignore
+			}
+		}
+    	return result;
+    }
+    
+    private static boolean createClusteredCacheFactoryStrategy() {
+    	try {
+    		PluginManager pluginManager = XMPPServer.getInstance().getPluginManager();
+    		String clusterPluginName = JiveGlobals.getProperty("cluster.plugin.name");
+    		Collection<Plugin> plugins = pluginManager.getPlugins();
+    		Plugin plugin = null;
+    		for (Plugin temp : plugins) {
+    			if(clusterPluginName.equals(pluginManager.getName(temp))) {
+    				plugin = temp;
+    				break;
+    			}
+			}
+    		
+    		if (plugin == null) {
+    			return false;
+    		}
+    		
+    		PluginClassLoader pluginClassloader = pluginManager.getPluginClassloader(plugin);
+    		
+    		clusteredCacheFactoryStrategy = (CacheFactoryStrategy) pluginClassloader
+    				.loadClass(clusteredCacheFactoryClass).newInstance();

```

Comment:
I would also recommend that the clusteredCacheFactoryClass be resolved via local scope at runtime, rather than defined as a static variable. The following can be removed from static scope and used here instead:
```
            String clusteredCacheFactoryClass = JiveGlobals.getProperty(CLUSTERED_CACHE_PROPERTY_NAME, 
                    "org.jivesoftware.openfire.plugin.util.cache.ClusteredCacheFactory");
```
A corresponding change can be made for the localCacheFactoryClass in the initialize() method.